modules/subchannel/src/problems/SubChannel1PhaseProblem.C

```diff
@@ -1906,217 +1953,152 @@ SubChannel1PhaseProblem::computeAddedHeatDuct(unsigned int i_ch, unsigned int iz
-  if (_monolithic_thermal_bool)
-  {
-    mat_array[Q * field_num + 3] = NULL;
-  }
-  LibmeshPetscCall(VecDuplicate(_amc_pressure_force_rhs, &vec_array[field_num]));
-  LibmeshPetscCall(VecCopy(_amc_pressure_force_rhs, vec_array[field_num]));
-  if (_pressure_axial_momentum_tight_coupling)
-  {
-    LibmeshPetscCall(VecAXPY(vec_array[field_num], 1.0, _amc_sys_mdot_rhs));
-  }
-  else
+  auto DupVecCopy = [&](Vec src, Vec * dst)
   {
-    unsigned int last_node = (iblock + 1) * _block_size;
-    unsigned int first_node = iblock * _block_size + 1;
-    LibmeshPetscCall(populateVectorFromHandle<SolutionHandle>(
-        _prod, *_mdot_soln, first_node, last_node, _n_channels));
-    Vec ls;
-    LibmeshPetscCall(VecDuplicate(_amc_sys_mdot_rhs, &ls));
-    LibmeshPetscCall(MatMult(_amc_sys_mdot_mat, _prod, ls));
-    LibmeshPetscCall(VecAXPY(ls, -1.0, _amc_sys_mdot_rhs));
-    LibmeshPetscCall(VecAXPY(vec_array[field_num], -1.0, ls));
-    LibmeshPetscCall(VecDestroy(&ls));
-  }
-
-  if (_verbose_subchannel)
-    _console << "Lin mom OK." << std::endl;
-
-  // Cross momentum conservation
-  field_num = 2;
-  mat_array[Q * field_num + 0] = NULL;
-  if (_pressure_cross_momentum_tight_coupling)
+    LibmeshPetscCall(VecDuplicate(src, dst));
+    LibmeshPetscCall(VecCopy(src, *dst));
+  };
+
+  const PetscInt Q = 3; // [mass conservation, axial momentum, cross momentum]
+
+  // small indexer
+  auto Idx = [&](PetscInt r, PetscInt c) { return Q * r + c; };
+
+  // arrays that MUST be declared before lambdas use them
+  std::vector<Mat> mat_array(Q * Q, NULL);
+  std::vector<Vec> vec_array(Q, NULL);
+
+  // generic assembler for one governing equation row in the nested matrix
+  auto AssembleEquation = [&](PetscInt f,
+                              Mat A0,
+                              Mat A1,
+                              Mat A2,             // three blocks in row f (can be nullptr)
+                              Vec rhs,            // base RHS for field f
+                              Vec rhs_add,        // optional extra RHS to add (can be nullptr)
+                              const char * label) // e.g. "Mass", "Lin mom", "Cross mom"
   {
-    LibmeshPetscCall(
-        MatDuplicate(_cmc_pressure_force_mat, MAT_COPY_VALUES, &mat_array[Q * field_num + 1]));
-    LibmeshPetscCall(MatAssemblyBegin(mat_array[Q * field_num + 1], MAT_FINAL_ASSEMBLY));
-    LibmeshPetscCall(MatAssemblyEnd(mat_array[Q * field_num + 1], MAT_FINAL_ASSEMBLY));
-  }
-  else
+    DupMatAssembled(A0, &mat_array[Idx(f, 0)]);
+    DupMatAssembled(A1, &mat_array[Idx(f, 1)]);
+    DupMatAssembled(A2, &mat_array[Idx(f, 2)]);
+    DupVecCopy(rhs, &vec_array[f]);
+    if (rhs_add)
+      LibmeshPetscCall(VecAXPY(vec_array[f], 1.0, rhs_add));
+    V(std::string(label) + " system assembled");
+  };
+
+  // -----------------------------------------------------------------------------
+  // Helper lambda that applies per-field under-relaxation by modifying BOTH the
```

Comment:
```suggestion
  // Helper lambda that applies per-equation under-relaxation by modifying BOTH the
```
a field is a variable. which there is a dedicated equation for, but still it's more about the equation than the field